### PR TITLE
Smart cropping

### LINKED ISSUE #54: Smart cropping options
Proposing 3 new values that can be set as the `crop` argument:

- `?crop=attention`: uses sharp's built in `attention` cropping strategy
- `?crop=entropy`: uses sharp's built in Shannon `entropy` cropping strategy
- `?crop=smart`: use the [smartcrop](https://github.com/jwagner/smartcrop-sharp) library - this might make a good default setting too.

How does the above sound @rmccue ?

### PR DESCRIPTION
**Pull request by roborourke**
- Adds support for sharp's built in attention & entropy cropping strategies
- Adds smartcrop-sharp lib

Usage: `?resize=300,300&crop_strategy=(smart|attention|entropy)`

Fixes #54

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -38,153 +39,179 @@ module.exports.s3 = function(config, key, args, callback) {
 	);
 };
 
+var getDimArray = function( dims ) {
+	return ( typeof dims === 'string'
+		? dims.split(',')
+		: dims
+	).map(function(v) {
```
Comment (rmccue) [inline index.js]: Oof, can we split this out to two statements instead? `var dimArr = ... ? ... : ... ;`

File: index.js [hunk 2]
```diff
@@ -38,153 +39,179 @@ module.exports.s3 = function(config, key, args, callback) {
...
+		return Number(v) || null;
+	});
+}
+
 module.exports.resizeBuffer = function(buffer, args, callback) {
+
+
```
Comment (rmccue) [inline index.js]: Bonus whitespace.

File: index.js [hunk 3]
```diff
@@ -38,153 +39,179 @@ module.exports.s3 = function(config, key, args, callback) {
...
-						})
-					);
-
-					// default to a black background to replicate Photon API behaviour
-					// when no background colour specified
-					if (!args.background) {
-						args.background = 'black';
+
+					// crop (assumes crop data from original)
+					if (args.crop && !args.crop.match(/(smart|attention|entropy)/)) {
```
Comment (rmccue) [inline index.js]: The regex needs to be anchored with `^` and `$`

File: index.js [hunk 4]
```diff
@@ -38,153 +39,179 @@ module.exports.s3 = function(config, key, args, callback) {
...
-						Number(args.h) || null
-					);
-					if (!args.crop) {
+
+					// resize
+					if (args.resize) {
+						args.resize = getDimArray( args.resize );
+
+						// apply cropping strategies
+						if ( args.crop === 'attention' ) {
```
Comment (rmccue) [inline index.js]: Seems strange to use the `resize` param for cropping? Right now, so far as I can see, they're two independent operations, so you could e.g. crop an image down to a 200x200 square then resize to 64x64.

Comment (roborourke) [reply index.js]: Resizing doesn't preserve aspect ratio so causes cropping by default in sharp. It's the param we use in the WP plugin to generate image sizes that have `crop` set to true so I wasn't really thinking of it that way.

I think it's much harder to figure out a crop size and then resize personally.

**Comment by roborourke** [comment]
@rmccue I don't fully agree here for a couple reasons:

- cropping occurs by default when using `resize` so it's not really separate
- if you have to crop first and then resize you've got two size calculations to figure out

I'd probably go with having the separate argument for `crop_strategy` that can be used alongside `resize` out the above options.

**Comment by roborourke** [comment]
Also worth noting that you can still crop and then resize with the approach I'm suggesting which is how it works now - automatic cropping only affects the behaviour of resizing when it has to crop the image too. That's how sharp works: http://sharp.pixelplumbing.com/en/stable/api-resize/#crop

It only applies a crop strategy or gravity when resizing.

File: index.js [hunk 5]
```diff
@@ -38,153 +39,178 @@ module.exports.s3 = function(config, key, args, callback) {
...
-						args.background = 'black';
+
+					// crop (assumes crop data from original)
+					if (args.crop) {
+						var cropValues =
+							typeof args.crop === 'string'
+								? args.crop.split(',')
+								: args.crop;
+
+						// convert percentages to px values
```
Comment (rmccue) [inline index.js]: 😍

File: index.js [hunk 6]
```diff
@@ -38,153 +39,178 @@ module.exports.s3 = function(config, key, args, callback) {
 	);
 };
 
+var getDimArray = function( dims ) {
+	var dimArr = ( typeof dims === 'string'
+		? dims.split(',')
+		: dims
+	);
```
Comment (rmccue) [inline index.js]: Can this just be a one-liner now? Seems strange to split out if we don't need the trailing `.map`. Also, the paren isn't needed if we do that.

File: index.js [hunk 7]
```diff
@@ -38,153 +39,178 @@ module.exports.s3 = function(config, key, args, callback) {
...
+	return dimArr.map(function(v) {
+		return Number(v) || null;
```
Comment (rmccue) [inline index.js]: Also, can this just `.map( Number )` instead? What's the implication of `NaN` vs `null`?

Comment (roborourke) [reply index.js]: Yep, `null` or `undefined` to use an automatic value with sharp. http://sharp.pixelplumbing.com/en/stable/api-resize/#parameters

**Pull request merged**